### PR TITLE
populate object name for admission attributes when CREATE

### LINKED ISSUE #46107: Admission does not populate object name on create
The admission attributes given to admission plugins do not populate object name on create.

This means that admission plugins must do `attr.GetObject().(*api.MyType).Name` themselves.

Since the object is available, the name in the admission attributes should be populated if available.

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
The admission attributes given to admission plugins do not populate object name on create.

Since the object is available, the name in the admission attributes should be populated if available.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #46107

**Special notes for your reviewer**:

**Release note**:

```release-note
populate object name for admission attributes when CREATE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/attributes.go [hunk 1]
```diff
@@ -57,6 +58,13 @@ func (record *attributesRecord) GetNamespace() string {
 }
 
 func (record *attributesRecord) GetName() string {
+	// On create, get name from new object if unset in admission
+	if len(record.name) == 0 && record.GetOperation() == Create {
+		metadata, err := meta.Accessor(record.GetObject())
+		if err == nil {
+			return metadata.GetName()
+		}
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/admission/attributes.go]: nit: log the error

**Comment by tallclair** [comment]
Looks reasonable to me.

File: staging/src/k8s.io/apiserver/pkg/admission/attributes.go [hunk 2]
```diff
@@ -57,6 +58,13 @@ func (record *attributesRecord) GetNamespace() string {
 }
 
 func (record *attributesRecord) GetName() string {
+	// On create, get name from new object if unset in admission
+	if len(record.name) == 0 && record.GetOperation() == Create {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/attributes.go]: I'd rather see this logic done where the record is constructed in the create rest handler... this object should remain really basic

File: staging/src/k8s.io/apiserver/pkg/admission/attributes.go [hunk 3]
```diff
@@ -57,6 +58,13 @@ func (record *attributesRecord) GetNamespace() string {
 }
 
 func (record *attributesRecord) GetName() string {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/attributes.go]: @deads2k I know you had thoughts about ordering of admission relative to generated name, which this does not address

File: staging/src/k8s.io/apiserver/pkg/admission/attributes.go (hunk 2 above)
Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/admission/attributes.go]: @liggitt Good suggestion.

**Comment by dixudx** [comment]
@liggitt @tallclair Updated. PTAL. Thanks.

**Comment by deads2k** [comment]
> @deads2k I know you had thoughts about ordering of admission relative to generated name, which this does not address

Yeah, I think we want to generate names ahead of admission.

File: staging/src/k8s.io/apiserver/pkg/admission/attributes.go [hunk 4]
```diff
@@ -35,6 +38,15 @@ type attributesRecord struct {
 }
 
 func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info) Attributes {
+	// On create, get name from new object if unset in admission
+	if len(name) == 0 && operation == Create {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/attributes.go]: hoist this logic even higher, to `rest.go#createHandler`

**Comment by dixudx** [comment]
@liggitt @tallclair Updated. PTAL. Thanks.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 5]
```diff
@@ -96,6 +98,15 @@ func createHandler(r rest.NamedCreater, scope RequestScope, typer runtime.Object
 		if admit != nil && admit.Handles(admission.Create) {
 			userInfo, _ := request.UserFrom(ctx)
 
+			// On create, get name from new object if unset
+			if len(name) == 0 {
+				metadata, err := meta.Accessor(obj)
+				if err == nil {
+					name = metadata.GetName()
+				} else {
+					glog.Errorf("unable to get name from obj %#v with err %v", obj, err)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: we don't glog here... return an error or do utilruntime.HandleError

**Comment by dixudx** [comment]
@liggitt Updated. PTAL.

**Comment by dixudx** [comment]
ping @liggitt @tallclair 
Rebased. PTAL. Thanks.

**Comment by tallclair** [comment]
> Yeah, I think we want to generate names ahead of admission.

@deads2k Can this be done in the defaulting stage? Then it would be compatible with the changes here.

**Comment by deads2k** [comment]
> @deads2k Can this be done in the defaulting stage? Then it would be compatible with the changes here.

It can't be done during API kind defaulting and because it is done in `BeforeCreate`, the name is known after `.Admit` (mutation) and before the `Validate` stage of admission.  Some refactoring would is required to do that part of `BeforeCreate` before admission is called.  I think the objectmeta pieces could be safely moved.

**Comment by dixudx** [comment]
ping @liggitt @tallclair Updated. PTAL. Thanks.

**Comment by tallclair** [comment]
I defer to @deads2k and @liggitt

**Comment by dixudx** [comment]
ping @deads2k @liggitt PTAL

I think this pr can be safely merged to v1.9.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 6]
```diff
@@ -94,6 +94,14 @@ func createHandler(r rest.NamedCreater, scope RequestScope, typer runtime.Object
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
+		// On create, get name from new object if unset
+		if len(name) == 0 {
+			if metadata, err := meta.Accessor(obj); err == nil {
+				name = metadata.GetName()
+			} else {
+				trace.Step(fmt.Sprintf("object %#v has no meta: %v", obj, err))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: This seems like an unusual use of tracing

**Review by liggitt — COMMENTED** [review]
One comment on the use of trace. I don't think this is critical for 1.9

**Comment by dixudx** [comment]
ping @liggitt Rebased and updated. PTAL. Thanks.

**Comment by dixudx** [comment]
ping @deads2k @liggitt PTAL. Thanks.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 7]
```diff
@@ -94,6 +96,14 @@ func createHandler(r rest.NamedCreater, scope RequestScope, typer runtime.Object
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
+		// On create, get name from new object if unset
+		if len(name) == 0 {
+			if metadata, err := meta.Accessor(obj); err == nil {
+				name = metadata.GetName()
+			} else {
+				glog.Warningf("object %#v has no meta: %v", obj, err)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: A failure here changes the invariant to the admission.  You need to fail, right?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 8]
```diff
@@ -94,6 +96,14 @@ func createHandler(r rest.NamedCreater, scope RequestScope, typer runtime.Object
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
+		// On create, get name from new object if unset
+		if len(name) == 0 {
+			if metadata, err := meta.Accessor(obj); err == nil {
+				name = metadata.GetName()
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: When are names generated.  Seems like we want to generated them *before* admission is checked.

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: Right. This did populate object name when `CREATE` before `admission.Admit` is called.

**Comment by dixudx** [comment]
Rebased. ping @deads2k and @liggitt for another look. Thanks.

**Comment by dixudx** [comment]
ping @deads2k @liggitt PTAL. Thanks.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go (hunk 8 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: > Right. This did populate object name when CREATE before admission.Admit is called.

You need some tests.  A few things look broken to me:
 1. if I included the name on create (block above), then choose a different name in the object, this will run admission on the name I claim, not the name I used to get past authorization.
 2. Generate name isn't done until the storage function is called (`r.Create` I think). That means this does not actually give the name being used.

Tests that would catch both conditions are needed.

**Comment by dixudx** [comment]
ping @deads2k @liggitt PTAL. Thanks.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go (hunk 8 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: I think both items from this comment are still outstanding.

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: @deads2k Did not find a good way to test this. Any suggestions?

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: @liggitt @deads2k Sorry for taking such a long time. Updated. PTAL. Thanks.

**Comment by liggitt** [comment]
from https://github.com/kubernetes/kubernetes/pull/53185#discussion_r164111736:

> You need some tests. A few things look broken to me:
>
> 1. if I included the name on create (block above), then choose a different name in the object, this will run admission on the name I claim, not the name I used to get past authorization.

Separately, we should sweep named create handlers to make sure they require the name in the object and the name in the URL to agree. These are the handlers that support named create:

```
/api/v1/namespaces/{namespace}/pods/{name}/binding
/api/v1/namespaces/{namespace}/pods/{name}/eviction
/apis/apps/v1beta1/namespaces/{namespace}/deployments/{name}/rollback
/apis/extensions/v1beta1/namespaces/{namespace}/deployments/{name}/rollback
```

> 2. Generate name isn't done until the storage function is called (`r.Create` I think). That means this does not actually give the name being used.

I'd be ok with deferring hoisting the generate name functionality out. I do think `AdmissionToValidateObjectFunc()` should pull the name out of the object if `staticAttributes.GetName()` is empty. That would let validating admission have access to the generated name.

> Tests that would catch both conditions are needed.

Agree.

**Comment by liggitt** [comment]
in AdmissionToValidateObjectFunc, for create operations, we should also pull the name from the incoming object if `staticAttributes.GetName()` is empty

by the time validating admission is called, we actually have the generated name populated

File: pkg/registry/core/pod/storage/eviction.go [hunk 9]
```diff
@@ -101,8 +101,15 @@ func propagateDryRun(eviction *policy.Eviction, options *metav1.CreateOptions) (
 }
 
 // Create attempts to create a new eviction.  That is, it tries to evict a pod.
-func (r *EvictionREST) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
-	eviction := obj.(*policy.Eviction)
+func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
+	eviction, ok := obj.(*policy.Eviction)
+	if !ok {
+		return nil, errors.NewBadRequest(fmt.Sprintf("not a Eviction object: %#v", obj))
```
Comment (liggitt) [inline pkg/registry/core/pod/storage/eviction.go]: use %T instead of %#v

File: pkg/registry/core/pod/storage/storage.go [hunk 10]
```diff
@@ -208,6 +219,32 @@ func (r *BindingREST) assignPod(ctx context.Context, podID string, machine strin
...
+// New creates a new binding resource
+func (r *LegacyBindingREST) New() runtime.Object {
+	return r.bindingRest.New()
+}
+
+// Create ensures a pod is bound to a specific host.
+func (r *LegacyBindingREST) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (out runtime.Object, err error) {
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, errors.NewBadRequest(fmt.Sprintf("not a Binding object: %#v", obj))
```
Comment (liggitt) [inline pkg/registry/core/pod/storage/storage.go]: %T

**Comment by liggitt** [comment]
drop this from the webhook admission test:
https://github.com/kubernetes/kubernetes/blob/19e85a9092399b065309aa5b4aec71a5b8101bab/test/integration/apiserver/admissionwebhook/admission_test.go#L298-L300

that will check that the name is wired into admission attributes correctly

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 11]
```diff
@@ -129,6 +130,15 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
+		// On create, get name from new object if unset
+		if len(name) == 0 {
+			_, name, err = scope.Namer.ObjectName(obj)
+			if err == nil {
+				if metadata, err := meta.Accessor(obj); err == nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: why is this modifying the object? I only expected to extract the name from the object

File: pkg/registry/core/pod/storage/eviction.go [hunk 12]
```diff
@@ -101,8 +101,15 @@ func propagateDryRun(eviction *policy.Eviction, options *metav1.CreateOptions) (
...
+	}
+
+	if name != eviction.Name {
```
Comment (liggitt) [inline pkg/registry/core/pod/storage/eviction.go]: `name` should be used in this method instead of `eviction.Name`

Comment (dixudx) [reply pkg/registry/core/pod/storage/eviction.go]: Here we are comparing whether `name ` equals that in the eviction object.

File: pkg/registry/core/pod/storage/eviction_test.go [hunk 13]
```diff
@@ -87,15 +76,20 @@ func TestEviction(t *testing.T) {
...
+		{
+			name: "matching pdbs with disruptions allowed, but bad name in Url",
+			pdbs: []runtime.Object{&policyv1beta1.PodDisruptionBudget{
+				ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "default"},
+				Spec:       policyv1beta1.PodDisruptionBudgetSpec{Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"a": "true"}}},
+				Status:     policyv1beta1.PodDisruptionBudgetStatus{PodDisruptionsAllowed: 1},
+			}},
+			badNameInURL: true,
+			eviction:     &policy.Eviction{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "default"}, DeleteOptions: metav1.NewDeleteOptions(0)},
+			expectError:  true,
```
Comment (liggitt) [inline pkg/registry/core/pod/storage/eviction_test.go]: verify this is the error you expect... the name you specify ("...badName")is not a valid name

File: pkg/registry/core/pod/storage/storage.go [hunk 14]
```diff
@@ -208,6 +219,32 @@ func (r *BindingREST) assignPod(ctx context.Context, podID string, machine strin
 	return
 }
 
+var _ = rest.Creater(&LegacyBindingREST{})
+
+// LegacyBindingREST implements the REST endpoint for binding pods to nodes when etcd is in use.
+type LegacyBindingREST struct {
```
Comment (liggitt) [inline pkg/registry/core/pod/storage/storage.go]: why was this needed? I thought BindingREST would be able to tell based on `len(name) == 0` whether it was being invoked as a subresource or a top-level resource

Comment (liggitt) [reply pkg/registry/core/pod/storage/storage.go]: actually, looks like the the name extraction in create.go will do the same thing this adapter is doing, and the name param will be populated... was there another reason LegacyBindingREST was created?

Comment (liggitt) [reply pkg/registry/core/pod/storage/storage.go]: mm... that triggers namedCreater, which triggers `includeName`, which returns an error if the scope.Namer.Name can't extract a name from the request

**Comment by liggitt** [comment]
had a few comments, and needs a rebase, but this looks close, thanks

good job keeping the binding/eviction changes in their own commits; we'll likely want to backport those

**Comment by liggitt** [comment]
unit test failure in TestCreateWithoutName looks legitimate

**Comment by dixudx** [comment]
Finally make all the CIs happy.

ping @liggitt for review. Thanks for your patient.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 15]
```diff
@@ -129,6 +116,19 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
+
+		_, name, err := scope.Namer.Name(req)
+		if err != nil {
+			if err.Error() != errEmptyName.Error() {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: `if err != errEmptyName`? it's odd to compare an interface's return value to a private error instance. other implementations of `Name()` wouldn't have a way to exercise this path

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 16]
```diff
@@ -129,6 +116,19 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
...
+				scope.err(err, w, req)
+				return
+			}
+
+			// On create, get name from new object if unset
+			if len(name) == 0 {
+				_, name, err = scope.Namer.ObjectName(obj)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: if this errors, the error is ignored

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 17]
```diff
@@ -43,7 +43,7 @@ import (
 	utiltrace "k8s.io/utils/trace"
 )
 
-func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Interface, includeName bool) http.HandlerFunc {
+func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Interface) http.HandlerFunc {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: dropping this loses information about whether the particular path expects name information in the path. I think this should be reverted.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 18]
```diff
@@ -57,20 +57,7 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		// TODO: we either want to remove timeout or document it (if we document, move timeout out of this function and declare it in api_installer)
 		timeout := parseTimeout(req.URL.Query().Get("timeout"))
 
-		var (
-			namespace, name string
-			err             error
-		)
-		if includeName {
-			namespace, name, err = scope.Namer.Name(req)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: revert this bit as well... if `includeName` is specified, we must get name info from the request

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: If reverting this, we need to keep this https://github.com/kubernetes/kubernetes/blob/61af419d386dd9aa05ddfd6e192981ae4ab0ffc5/plugin/pkg/admission/noderestriction/admission.go#L364-L367 in noderestriction plugin.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: hmm... the main thing I care about is keeping the `includeName=true` condition tightly connected to requiring scope.Namer.Name to succeed

what about this:
```
namespace, name, err := scope.Namer.Name(req)
if err != nil {
  if includeName {
    // name was required, return
    scope.err(err, w, req)
    return
  }

  // otherwise attempt to look up the namespace
  namespace, err = scope.Namer.Namespace(req)
  if err != nil {
      scope.err(err, w, req)
      return
  }
}
```

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: if that is satisfied, looking up the name from the object later if needed can be best effort:

```
if len(name) == 0 {
  _, name, _ = scope.Namer.ObjectName(obj)
}
```

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: > if includeName is specified, we must get name info from the request

https://github.com/kubernetes/kubernetes/blob/bb381d6e0af36b56835d7917831d736490b4e7a5/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L129-L132

This check has been added.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: moving it down below allows requests that expected to get name info from the request to instead get it from the object. https://github.com/kubernetes/kubernetes/pull/53185/files#r313210079 preserves the current requirement for `includeName` to be satisfied by the request

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 19]
```diff
@@ -129,6 +131,15 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
+
+		// On create, get name from new object if unset
+		if len(name) == 0 {
+			_, name, err = scope.Namer.ObjectName(obj)
+			if err != nil && includeName {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: if `includeName` is true, name cannot be length 0 (we would have errored above if that is the case)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 20]
```diff
@@ -129,6 +131,14 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
 
 		userInfo, _ := request.UserFrom(ctx)
+
+		// On create, get name from new object if unset
+		if len(name) == 0 {
+			if _, name, err = scope.Namer.ObjectName(obj); err != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: Not all submitted API objects contain names (for example, non-persisted review objects and objects using `GenerateName`)

We've already enforced `includeName` above, so ignore errors here.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: also, to handle the GeneratedName case, in AdmissionToValidateObjectFunc, if `staticAttributes.GetName()` is empty, shouldn't we attempt to override with obj.GetName()? A testcase with generate name would be good (mutating admission wouldn't currently get the name attribute, validating admission would)

**Comment by dixudx** [comment]
@liggitt Please take a review. Thanks.

**Comment by liggitt** [comment]
/priority important-soon
/kind feature

needs rebase

**Comment by dixudx** [comment]
@liggitt Rebased. PTAL. Thanks.
